ADD: 레벨, 적 클래스 지정, 스폰될 수 데이터테이블 관리 추가 및 코멘트 수정 처리

Jo-sm94 opened the pull request:
레벨, 적 클래스 지정, 스폰될 수를 데이터테이블로 관리 할 수 있도록 기능 추가
코멘트 수정 처리
 - 방어코드 (null 체크, 비정상 로직 발생 가능성 처리)
 - 레벨명 체크
 - 적 스폰 요청 수와 실제 스폰된 적 수 비교 처리
 - 접근 제어 처리(변수, 함수) (getter, setter)

--------------------------------------------------------------------
TODO
- 적 스폰되는 위치 조정 (Z축 지형에 따라 땅에서 시작할수있게)
 - 임시 Game Start 변경 필요
 - Game End 시 분기처리 부분
 - 전환 레벨(LoadStreamLevel)

Jo-sm94 committed:
ADD: 레벨, 적 클래스 지정, 스폰될 수 데이터테이블 관리 추가 및 코멘트 수정 처리

데이터테이블로 관리 할 수 있도록 기능 추가
코멘트 수정 처리
 - 방어코드 (null 체크, 비정상 로직 발생 가능성 처리)
 - 레벨명 체크
 - 적 스폰 요청 수와 실제 스폰된 적 수 비교 처리
 - 접근 제어 처리(변수, 함수) (getter, setter)

Jo-sm94 committed:
Merge branch 'Develop' into feature/GameMode

hwarang98 commented on Source/Shooter/Private/GameModes/ShooterGameStateBase.cpp:
```diff
@@ -15,10 +15,14 @@ void AShooterGameStateBase::SetAliveEnemyCount(int32 Count)
-	AliveEnemyCount--;
+	if (AliveEnemyCount > 0)
+	{
+		AliveEnemyCount--;
```
이 함수가 어디에서, 누가, 몇 번 호출하는지 완전히 제어할 수 없다면, 다음과 같은 경우가 생길수도있어요! 

1. 적이 두 번 죽음 처리되는 버그
   1-1. (예: 같은 적이 동시에 두 개의 타격을 받아 두 번 OnEnemyDied() 호출)

2. 네트워크 환경에서 중복 호출(네트워크가 없지만..!)

그래서 이런코드는 보통
`AliveEnemyCount = FMath::Max(AliveEnemyCount - 1, 0);` 이런 방식이 좋습니다!

hwarang98 commented on Source/Shooter/Private/GameModes/ShooterGameStateBase.cpp:
```diff
@@ -15,10 +15,14 @@ void AShooterGameStateBase::SetAliveEnemyCount(int32 Count)
-	if (AliveEnemyCount <= 0)
+	if (AliveEnemyCount <= 0 && TotalEnemy > 0)
 	{
 		AShooterBaseGameMode* GameMode = GetWorld()->GetAuthGameMode<AShooterBaseGameMode>();
```
해당 코드도 너무 잘짜신 코드지만 매번 gamemode를 가지고 오는건 그만큼 연산을 하는 작업이기때문에 해더파일에 `TWeakObjectPtr<AShooterBaseGameMode> CachedGameMode;` 를 선언후 생성자에서 할당후 해당 함수에서 사용하는 방법도 좋아보입니다! (최초 1회만 호출되도 되는경우)

hwarang98 commented on Source/Shooter/Private/GameModes/ShooterGameInstance.cpp:
```diff
@@ -3,30 +3,53 @@
-			WaveLevelNames.Add(FName(*LevelName));
+			if (Row)
+			{
+				WaveConfigs.Add(*Row);
```
해당 코드는 구조체를 복사합니다
FWaveConfig가 크거나 중첩된 배열/맵이 있다면 비용이 커요!
(같은 데이터를 다른 메모리에 할당하기때문에 같은 데이터가 다른 메모리에 할당하는거기때문에 불필요한 메모리할당이 이루어 질수도있습니다) 

`WaveConfigs.Add(MoveTemp(*Row));` 이런식으로 사용할수도 있습니다!

MoveTemp 함수는
```cpp
A의 메모리를 복사하지 않고 Array가 A의 자원을 '뺏어갑니다'
A는 이후 비워진 상태 (유효하지만, 값은 무의미)
마치 이사하면서 짐을 옮기고 빈 집이 남는 것과 비슷함
```
다만 주의할점은 movetemp로 이사후 이전에 쓰이던 객체는 비어버리므로 한번 검토후 적용 부탁드립니다!

hwarang98 commented on Source/Shooter/Private/GameModes/ShooterBaseGameMode.cpp:
```diff
@@ -18,14 +19,15 @@ void AShooterBaseGameMode::StartPlay()
-	if (CurrentMapName.Contains("MenuLevel"))
+	//FString CurrentMapName = GetWorld()->GetMapName();
+	FString CleanMapName = UGameplayStatics::GetCurrentLevelName(this, true);
+	if (CleanMapName.Contains("MenuLevel"))
```
맵 이름 비교 방식 개선 고려 해보면 좋을것 같아요
Contains()는 부분 일치라 `SubMenuLevel` 같은 다른 이름에도 반응할 수 있습니다! 

보다 정확하게 매치하고 싶다면 Equals() 사용도 검토 해보시면 좋을것 같아요!

ex) `if (CleanMapName.Equals(TEXT("MenuLevel")))`

hwarang98 commented on Source/Shooter/Private/GameModes/ShooterBaseGameMode.cpp:
```diff
@@ -44,51 +46,93 @@ void AShooterBaseGameMode::StartPlay()
+		return;
+	}
+
+	const FWaveConfig& WaveData = GameInstance->GetWaveConfigs()[GameInstance->GetCurrentWave()];
```
해당 waveData는 `	if (!GameInstance->GetWaveConfigs().IsValidIndex(CurrentWave - 1)) return;` 유효성 뒤에 접근하는게 더 자연스럽긴한데 혹시 다른 의도한 바 가있을까요?

Jo-sm94 replied:
구조체의 index 문제때문에 그랬던건데 유효성 뒤에 접근하는게 맞는거 같습니다.
웨이브는 1, 2, 3, 4  구조체 인덱스는 0, 1, 2, 3 이라서 웨이브에 1추가하기전에 선언해놨던겁니다.
선언부를 내려서 작업하겠습니다
```
if (!GameInstance->GetWaveConfigs().IsValidIndex(CurrentWave - 1)) return;

const FWaveConfig& WaveData = GameInstance->GetWaveConfigs()[GameInstance->GetCurrentWave() - 1];
const TArray<TSubclassOf<ACharacter>>& EnemyTypes = WaveData.EnemyClasses;
int32 SpawnedCount = 0;
```

hwarang98 commented on Source/Shooter/Private/GameModes/ShooterBaseGameMode.cpp:
```diff
@@ -44,51 +46,93 @@ void AShooterBaseGameMode::StartPlay()
+	const TArray<TSubclassOf<ACharacter>>& EnemyTypes = WaveData.EnemyClasses;
+	int32 SpawnedCount = 0;
+	
+	GameInstance->SetCurrentWave(GameInstance->GetCurrentWave()+1);
```
요것도 현재 웨이브를 가지고 오기전에 set을 해주는 코드로 해석되는데 `int CurrentWave = GameInstance->GetCurrentWave();` 로 먼저 wave정보를 가지고온후 set해주는게 자연스러워 보이는데 이부분 도 검토후 수정 부탁드릴께요~

Jo-sm94 replied:
아 이 부분을 고치면 위에서 코멘트 해주신 부분까지 깔끔하게 처리 가능할 것 같습니다.
감사합니다

hwarang98 replied:
좋습니다!!

hwarang98 commented on Source/Shooter/Public/GameModes/ShooterBaseGameMode.h:
```diff
@@ -12,19 +12,28 @@ class SHOOTER_API AShooterBaseGameMode : public AGameModeBase
+
 	AShooterBaseGameMode();
+
+	UFUNCTION(BlueprintCallable)
```
요기 Public도 protected나 private 으로 변경해주시는게 좋아보입니다(캡슐화)! 다른 헤더 파일도 public되어있는 부분도 한번 검토 부탁드릴께요!

Jo-sm94 committed:
ADD: 레벨, 적 클래스 지정, 스폰될 수 데이터테이블 관리 추가 및 코멘트 수정 처리#2

레벨, 적 클래스 지정, 스폰될 수를 데이터테이블로 관리 할 수 있도록 기능 추가
코멘트 수정 처리 #2
 - 방어코드 (null 체크, 비정상 로직 발생 가능성 처리)
 - 레벨명 체크
 - 적 스폰 요청 수와 실제 스폰된 적 수 비교 처리
 - 접근 제어 처리(변수, 함수) (getter, setter)
